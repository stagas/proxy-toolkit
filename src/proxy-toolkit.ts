export type GetterReturn<T> = { [K: string]: T }

/**
 * Creates a proxy that calls `cb` for every property accessed and returns its result.
 *
 * @param cb Called with the key of the property being accessed.
 */
export const Getter = <T>(cb: (key: string) => T): GetterReturn<T> =>
  new Proxy({}, { get: (_, key: string) => cb(key) })

export type FluentCapture = (readonly ['get', string | symbol] | readonly ['apply', any[]])[]

export const FluentCaptureSymbol = Symbol()

export const FluentCapture = () => {
  const fn = () => {}
  const results: FluentCapture = []
  const proxy: any = new Proxy(fn, {
    get: (_, key, receiver) => {
      if (key === FluentCaptureSymbol) return true
      if (key === '_results') return results
      results.push(['get', key] as const)
      return receiver
    },
    apply: (_, __, args: any[]) => {
      results.push(['apply', args] as const)
      return proxy
    },
  })
  return proxy
}

export const applyFluentCapture = (capture: FluentCapture, target: any) => {
  let res = target
  for (const [op, args] of capture) {
    if (op === 'get') {
      res = res[args]
    } else if (op === 'apply') {
      res = res(...args)
    }
  }
  return res
}
