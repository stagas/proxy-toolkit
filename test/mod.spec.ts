import { applyFluentCapture, FluentCapture, FluentCaptureSymbol, Getter } from '../src/proxy-toolkit'

describe('Getter(cb)', () => {
  it('creates a getter', () => {
    const results: string[] = []
    const get = Getter(key => 'hello' + key)
    results.push(get.foo)
    results.push(get.bar)
    expect(results).toEqual(['hellofoo', 'hellobar'])
  })
})

describe('FluentCapture(cb)', () => {
  it('creates a fluent capture', () => {
    const fluent = FluentCapture()
    expect(fluent[FluentCaptureSymbol]).toBe(true)

    fluent.some.thing(10).is.ok()
    expect(fluent._results).toMatchSnapshot()

    const other = FluentCapture()
    applyFluentCapture(fluent._results, other)
    expect(other._results).toEqual(fluent._results)
  })
})
