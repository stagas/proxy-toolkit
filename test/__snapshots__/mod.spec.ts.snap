
exports[`FluentCapture(cb) creates a fluent capture 1`] = `
Array [
  Array [
    "get",
    "some",
  ],
  Array [
    "get",
    "thing",
  ],
  Array [
    "apply",
    Array [
      10,
    ],
  ],
  Array [
    "get",
    "is",
  ],
  Array [
    "get",
    "ok",
  ],
  Array [
    "apply",
    Array [],
  ],
]
`;
